python/az/aro/azext_aro/_dynamic_validators.py

```diff
@@ -0,0 +1,279 @@
+                f"Invalid -- subnet, error when getting '{subnet}': {str(err)}") from err
+        raise CLIInternalError(
+            f"Unexpected error when getting subnet '{subnet}': {str(err)}") from err
+
+    return subnet_obj
+
+
+def dyn_validate_vnet(key):
+    def _validate_vnet(cmd, namespace):
+        errors = []
+        with AutomaticSpinner("Validating Virtual Network Permissions"):
+            vnet = getattr(namespace, key)
+
+            if not is_valid_resource_id(vnet):
+                raise RequiredArgumentMissingError(
+                    f"Must specify --vnet if --{key.replace('_', '-')} is not an id.")
+
+            validate_vnet(cmd, namespace)
+
+            parts = parse_resource_id(vnet)
+
+            network_client = get_mgmt_service_client(
+                cmd.cli_ctx, ResourceType.MGMT_NETWORK)
+
+            auth_client = get_mgmt_service_client(
+                cmd.cli_ctx, ResourceType.MGMT_AUTHORIZATION, api_version="2015-07-01")
+
+            try:
+                network_client.virtual_networks.get(parts['resource_group'], parts['name'])
+            except Exception as err:
+                if isinstance(err, ResourceNotFoundError):
+                    raise InvalidArgumentValueError(
+                        f"Invalid --{key.replace('_', '-')}, error when getting '{vnet}': {str(err)}") from err
+                raise CLIInternalError(
+                    f"Unexpected error when getting subnet '{vnet}': {str(err)}") from err
+
+            errors = validate_resource(auth_client, key, parts, [
+                "Microsoft.Network/virtualNetworks/join/action",
+                "Microsoft.Network/virtualNetworks/read",
+                "Microsoft.Network/virtualNetworks/write",
+                "Microsoft.Network/virtualNetworks/subnets/join/action",
+                "Microsoft.Network/virtualNetworks/subnets/read",
+                "Microsoft.Network/virtualNetworks/subnets/write", ])
+
+        return errors
+
+    return _validate_vnet
+
+
+def dyn_validate_subnet(key):
+    def _validate_subnet(cmd, namespace):
+        errors = []
+        with AutomaticSpinner(f"Validating {key} permissions"):
+            subnet = getattr(namespace, key)
+
+            if not is_valid_resource_id(subnet):
+                if not namespace.vnet:
+                    raise RequiredArgumentMissingError(
+                        f"Must specify --vnet if --{key.replace('_', '-')} is not an id.")
+
+                validate_vnet(cmd, namespace)
+
+                subnet = namespace.vnet + '/subnets/' + subnet
+                setattr(namespace, key, subnet)
+
+            parts = parse_resource_id(subnet)
+
+            network_client = get_mgmt_service_client(
+                cmd.cli_ctx, ResourceType.MGMT_NETWORK)
+
+            auth_client = get_mgmt_service_client(
+                cmd.cli_ctx, ResourceType.MGMT_AUTHORIZATION, api_version="2015-07-01")
+
+            try:
```

Comment:
Is it possible to also check at this time if the subnet has an NSG already attached? And if so, should we?

I know we currently run into a race if a newly created VNET/subnet is used when creating a cluster the RP can lose the race to azure and azure applies its own NSG before the RP can.

Reply:
I'm actually not too sure on this. Would this be a check that is needed in order to successfully create the cluster? @darthhexx Thoughts?

Reply:
I haven't come across this in my tests before, but if we didn't have the ACLs to replace the Azure added NSG with the RP NSG, then this could well be an issue. Let's sync on this @ellis-johnson and get some test cases together to see if we can deterministically test for this edge case.

Reply:
I discussed with @darthhexx and decided to add a check in to see if the subnets are already assigned to an NSG